aws/rust-runtime/aws-config/src/sts/assume_role.rs

```diff
@@ -0,0 +1,206 @@
+    }
+}
+
+/// A builder for [`AssumeRoleProvider`].
+///
+/// Construct one through [`AssumeRoleProvider::builder`].
+pub struct AssumeRoleProviderBuilder {
+    role_arn: String,
+    external_id: Option<String>,
+    session_name: Option<String>,
+    region: Option<Region>,
+    connection: Option<smithy_client::erase::DynConnector>,
+}
+
+impl AssumeRoleProviderBuilder {
+    /// Start a new assume role builder for the given role.
+    ///
+    /// The `role` argument should take the form an Amazon Resource Name (ARN) like
+    ///
+    /// ```text
+    /// arn:aws:iam::123456789012:role/example
+    /// ```
+    pub fn new(role: impl Into<String>) -> Self {
+        Self {
+            role_arn: role.into(),
+            external_id: None,
+            session_name: None,
+            region: None,
+            connection: None,
+        }
+    }
+
+    /// Set a unique identifier that might be required when you assume a role in another account.
+    ///
+    /// If the administrator of the account to which the role belongs provided you with an external
+    /// ID, then provide that value in this parameter. The value can be any string, such as a
+    /// passphrase or account number.
+    pub fn external_id(mut self, id: impl Into<String>) -> Self {
+        self.external_id = Some(id.into());
+        self
+    }
+
+    /// Set an identifier for the assumed role session.
+    ///
+    /// Use the role session name to uniquely identify a session when the same role is assumed by
+    /// different principals or for different reasons. In cross-account scenarios, the role session
+    /// name is visible to, and can be logged by the account that owns the role. The role session
+    /// name is also used in the ARN of the assumed role principal.
+    pub fn session_name(mut self, name: impl Into<String>) -> Self {
+        self.session_name = Some(name.into());
+        self
+    }
+
+    /// Set the region to assume the role in.
+    ///
+    /// This dictates which STS endpoint the AssumeRole action is invoked on.
+    pub fn region(mut self, region: Region) -> Self {
+        self.region = Some(region);
+        self
+    }
+
+    /// Set the backing connection to use when talking to STS.
+    pub fn connection(mut self, conn: impl smithy_client::bounds::SmithyConnector) -> Self {
+        self.connection = Some(smithy_client::erase::DynConnector::new(conn));
+        self
+    }
+
+    /// Build a credentials provider for this role authorized by the given `provider`.
+    pub fn build(self, provider: impl Into<SharedCredentialsProvider>) -> AssumeRoleProvider {
+        let config = aws_sdk_sts::Config::builder()
+            .credentials_provider(provider.into())
+            .region(self.region.clone())
+            .build();
+
+        let client = if let Some(conn) = self.connection {
+            aws_hyper::Client::new(conn)
+        } else {
+            aws_hyper::Client::https()
+        };
```

Comment:
```suggestion
let conn = self.connection.unwrap_or(crate::expect_connector(crate::default_connector());
let client = aws_hyper::Client::new(conn);
```

this will use the client as configured via crate features & give a consistent error message if all the client implementations have been disabled

Reply:
I assume you mean `unwrap_or_else` so that we don't create the default connection every time unnecessarily?